lib/CL/devices/level0/pocl-level0.cc

```diff
@@ -0,0 +1,1227 @@
+        }
+        case OCLType::Pointer: {
+          meta->arg_info[j].type = POCL_ARG_TYPE_POINTER;
+          meta->arg_info[j].type_size = sizeof(cl_mem);
+          switch (FI->ArgTypeInfo[j].Space) {
+          case OCLSpace::Private:
+            Addr = CL_KERNEL_ARG_ADDRESS_PRIVATE;
+            break;
+          case OCLSpace::Local:
+            Addr = CL_KERNEL_ARG_ADDRESS_LOCAL;
+            break;
+          case OCLSpace::Global:
+            Addr = CL_KERNEL_ARG_ADDRESS_GLOBAL;
+            break;
+          case OCLSpace::Constant:
+            Addr = CL_KERNEL_ARG_ADDRESS_CONSTANT;
+            break;
+          case OCLSpace::Unknown:
+            Addr = CL_KERNEL_ARG_ADDRESS_PRIVATE;
+            break;
+          }
+          break;
+        }
+        case OCLType::Image: {
+          meta->arg_info[j].type = POCL_ARG_TYPE_IMAGE;
+          meta->arg_info[j].type_size = sizeof(cl_mem);
+          Addr = CL_KERNEL_ARG_ADDRESS_GLOBAL;
+          bool Readable = FI->ArgTypeInfo[j].Attrs.ReadableImg;
+          bool Writable = FI->ArgTypeInfo[j].Attrs.WriteableImg;
+          if (Readable && Writable)
+            Access = CL_KERNEL_ARG_ACCESS_READ_WRITE;
+          if (Readable && !Writable)
+            Access = CL_KERNEL_ARG_ACCESS_READ_ONLY;
+          if (!Readable && Writable)
+            Access = CL_KERNEL_ARG_ACCESS_WRITE_ONLY;
+          break;
+        }
+        case OCLType::Sampler: {
+          meta->arg_info[j].type = POCL_ARG_TYPE_SAMPLER;
+          meta->arg_info[j].type_size = sizeof(cl_mem);
+          break;
+        }
+        case OCLType::Opaque: {
+          POCL_MSG_ERR("Unknown OCL type OPaque\n");
+          meta->arg_info[j].type = POCL_ARG_TYPE_NONE;
+          meta->arg_info[j].type_size = FI->ArgTypeInfo[j].Size;
+          break;
+        }
+        }
+        meta->arg_info[j].address_qualifier = Addr;
+        meta->arg_info[j].access_qualifier = Access;
+        meta->arg_info[j].type_qualifier = CL_KERNEL_ARG_TYPE_NONE;
+        if (FI->ArgTypeInfo[j].Attrs.Constant)
+          meta->arg_info[j].type_qualifier = CL_KERNEL_ARG_TYPE_CONST;
+        if (FI->ArgTypeInfo[j].Attrs.Restrict)
+          meta->arg_info[j].type_qualifier = CL_KERNEL_ARG_TYPE_RESTRICT;
+        if (FI->ArgTypeInfo[j].Attrs.Volatile)
+          meta->arg_info[j].type_qualifier = CL_KERNEL_ARG_TYPE_VOLATILE;
+      }
+
+      // TODO: POCL_HAS_KERNEL_ARG_TYPE_NAME missing
+      meta->has_arg_metadata = POCL_HAS_KERNEL_ARG_ACCESS_QUALIFIER |
+                               POCL_HAS_KERNEL_ARG_ADDRESS_QUALIFIER |
+                               POCL_HAS_KERNEL_ARG_TYPE_QUALIFIER |
+                               POCL_HAS_KERNEL_ARG_NAME;
+    }
+
+    // meta->num_locals
+    // meta->local_sizes
+    // meta->req_wg_size
+    // meta->wg_size_hint
+    // meta->vec_type_hint
+    // meta->attributes
+    ++i;
+  }
+
+  return 1;
+}
+
+int pocl_level0_create_kernel(cl_device_id device, cl_program program,
```

Comment:
This is a C++ file so we have to adhere to the C++ guidelines with the new codes, meaning initial cap variables, C++ comments etc.

Reply:
fixed